larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+
+	  if (endpointDistance < currentEndpointDistance)
+	    {    
+	      currentEndpointDistance = endpointDistance; 
+	      bestClusterList.clear();
+	      bestClusterList.push_back(pCluster);
+	    }    
+	} 
+
+      if (bestClusterList.size() == 0)
+	{
+	  throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+	}
+
+      const Cluster *const pCluster(*(bestClusterList.begin())); 
+      return pCluster;
+
+    }
+    else{
+      throw StatusCodeException(STATUS_CODE_FAILURE);
+      return 0;
+    }
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::SetEndpoints(DirectionFitObject &fitResult, const Cluster *const pCluster)
+{
+    CartesianVector lowZVector(0.f, 0.f, 0.f), highZVector(0.f, 0.f, 0.f);
+    LArClusterHelper::GetExtremalCoordinates(pCluster, lowZVector, highZVector);
+
+    fitResult.SetBeginpoint(lowZVector);
+    fitResult.SetEndpoint(highZVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::SetEndpoints(DirectionFitObject &fitResult, const LArTrackStateVector &trackStateVector)
+{
+    TrackState firstTrackState(*(trackStateVector.begin())), lastTrackState(trackStateVector.back());
+    const pandora::CartesianVector initialPosition(firstTrackState.GetPosition());
+    const pandora::CartesianVector endPosition(lastTrackState.GetPosition());
+
+    const pandora::CartesianVector lowZVector(initialPosition.GetZ() < endPosition.GetZ() ? initialPosition : endPosition);
+    const pandora::CartesianVector highZVector(initialPosition.GetZ() > endPosition.GetZ() ? initialPosition : endPosition);
+
+    fitResult.SetBeginpoint(lowZVector);
+    fitResult.SetEndpoint(highZVector);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::FillHitChargeVector(const Cluster *const pCluster, HitChargeVector &hitChargeVector)
+{
+    OrderedCaloHitList orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    CaloHitList caloHitList;
+    orderedCaloHitList.FillCaloHitList(caloHitList);
+
+    const TwoDSlidingFitResult &slidingFit(this->GetCachedSlidingFit(pCluster));
+
+    for (CaloHitList::const_iterator hitIter = caloHitList.begin(), hitIterEnd = caloHitList.end(); hitIter != hitIterEnd; ++hitIter)
+    {
+        const CaloHit *const pCaloHit(*hitIter);
+        const CartesianVector caloHitPosition(pCaloHit->GetPositionVector());
+        float hitWidth(pCaloHit->GetCellSize1());
+
+        float caloHitEnergy(pCaloHit->GetInputEnergy());
+        caloHitEnergy *= 273.5; //ADC to electron
+        caloHitEnergy *= 23.6/1000000; //ionisation energy per electron in MeV
+        caloHitEnergy /= 0.62;
```

Comment:
Won't these be standard for the conversion though?

Reply:
I would imagine ADC to electron has some detector dependency and I'm not sure what 0.62 represents here. To what extent each parameter is a configurable can be determined, but as a starting point I would at least provide some named constants to help the interpretation of the operations, e.g. const float ioniEnergyPerElectron{2.36e-5} (It's better to have self-documenting code rather than comments where this is possible).

Reply:
Right okay, done